File: backend/src/main/java/com/edunexus/model/Course.java
package com.edunexus.model;

import lombok.Data;
import org.springframework.data.annotation.Id;
import org.springframework.data.mongodb.core.mapping.Document;

import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

@Data
@Document(collection = "courses")
public class Course {
    @Id
    private String id;
    
    private String title;
    private String description;
    private String adminId;
    private Set<String> enrolledStudents = new HashSet<>();
    private Set<String> teachingAssistants = new HashSet<>();
    
    private List<CourseModule> modules = new ArrayList<>();
    private List<Announcement> announcements = new ArrayList<>();
    private List<Discussion> discussions = new ArrayList<>();
    
    private LocalDateTime createdAt;
    private LocalDateTime updatedAt;
    
    @Data
    public static class CourseModule {
        private String id;
        private String title;
        private String description;
        private List<ModuleContent> contents = new ArrayList<>();
        private LocalDateTime dueDate;
    }
    
    @Data
    public static class ModuleContent {
        private String id;
        private String title;
        private ContentType type;
        private String content; // URL or text content
        private boolean required;
        
        public enum ContentType {
            VIDEO,
            DOCUMENT,
            ASSIGNMENT,
            QUIZ,
            REFERENCE_BOOK
        }
    }
    
    @Data
    public static class Announcement {
        private String id;
        private String title;
        private String content;
        private String authorId;
        private LocalDateTime createdAt;
    }
    
    @Data
    public static class Discussion {
        private String id;
        private String title;
        private String content;
        private String authorId;
        private boolean isAnonymous;
        private List<DiscussionReply> replies = new ArrayList<>();
        private LocalDateTime createdAt;
    }
    
    @Data
    public static class DiscussionReply {
        private String id;
        private String content;
        private String authorId;
        private boolean isAnonymous;
        private LocalDateTime createdAt;
    }
} 

File: backend/src/models/Course.ts
import mongoose from 'mongoose';

const moduleContentSchema = new mongoose.Schema({
  title: {
    type: String,
    required: true,
  },
  type: {
    type: String,
    enum: ['VIDEO', 'DOCUMENT', 'ASSIGNMENT', 'QUIZ', 'REFERENCE_BOOK'],
    required: true,
  },
  content: {
    type: String,
    required: true,
  },
  required: {
    type: Boolean,
    default: true,
  },
});

const courseModuleSchema = new mongoose.Schema({
  title: {
    type: String,
    required: true,
  },
  description: {
    type: String,
    required: true,
  },
  contents: [moduleContentSchema],
  dueDate: {
    type: Date,
    required: true,
  },
});

const announcementSchema = new mongoose.Schema({
  title: {
    type: String,
    required: true,
  },
  content: {
    type: String,
    required: true,
  },
  authorId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: true,
  },
  createdAt: {
    type: Date,
    default: Date.now,
  },
});

const discussionReplySchema = new mongoose.Schema({
  content: {
    type: String,
    required: true,
  },
  authorId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: true,
  },
  isAnonymous: {
    type: Boolean,
    default: false,
  },
  createdAt: {
    type: Date,
    default: Date.now,
  },
});

const discussionSchema = new mongoose.Schema({
  title: {
    type: String,
    required: true,
  },
  content: {
    type: String,
    required: true,
  },
  authorId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: true,
  },
  isAnonymous: {
    type: Boolean,
    default: false,
  },
  replies: [discussionReplySchema],
  createdAt: {
    type: Date,
    default: Date.now,
  },
});

const courseSchema = new mongoose.Schema({
  title: {
    type: String,
    required: true,
  },
  description: {
    type: String,
    required: true,
  },
  adminId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: true,
  },
  enrolledStudents: [{
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
  }],
  teachingAssistants: [{
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
  }],
  modules: [courseModuleSchema],
  announcements: [announcementSchema],
  discussions: [discussionSchema],
  createdAt: {
    type: Date,
    default: Date.now,
  },
  updatedAt: {
    type: Date,
    default: Date.now,
  },
});

export default mongoose.model('Course', courseSchema); 